extensions/cornerstone/src/init.tsx

```diff
@@ -182,9 +183,44 @@ export default async function init({
     commandsManager,
   });
 
+  /**
+   * When a viewport gets a new display set, this call will go through all the
+   * active tools in the toolbar, and call any commands registered in the
+   * toolbar service with a callback to re-enable on displaying the viewport.
+   */
   const newStackCallback = evt => {
     const { element } = evt.detail;
-    utilities.stackPrefetch.enable(element);
+    const activeTools = toolbarService.getActiveTools();
+
+    activeTools.forEach(tool => {
+      const toolData = toolbarService.getNestedButton(tool);
+      const commands = toolData?.listeners?.newStack;
+      commandsManager.run(commands, { element, evt });
+    });
+  };
+
+  const activeViewportIdChangedListener = evt => {
+    const { viewportId } = evt;
+    const toolGroup = toolGroupService.getToolGroupForViewport(viewportId);
+
+    const activeTools = toolbarService.getActiveTools();
+
+    activeTools.forEach(tool => {
+      if (!toolGroup?._toolInstances?.[tool]) {
+        return;
+      }
+
+      // check if tool is active on the new viewport
+      const toolEnabled = toolGroup._toolInstances[tool].mode === Enums.ToolModes.Enabled;
+
+      if (!toolEnabled) {
+        return;
+      }
+
+      const toolData = toolbarService.getNestedButton(tool);
+      const commands = toolData?.listeners?.activeViewportIdChanged;
+      commandsManager.run(commands, { viewportId, evt });
```

Comment:
this seems to be a common pattern, maybe we refactor? and maybe rename toolData to buttons

Reply:
Renamed but didn't factor out the common pattern.  Will think about what is worth doing there once we have one or two more examples.